Introduce multistore technical documentation for configuration forms and settings (#843)
PrestaShop/docs: src/content/1.7/development/Multistore/configuration-forms/_index.md

---
title: Configuration forms
menuTitle: Configuration forms
summary: "Configuration forms when multistore is enabled"
weight: 50
---

# Configuration forms in multistore context
{{< minver v="1.7.8" title="true" >}}

> Since Prestashop 1.7.8, when multistore is enabled, migrated configuration forms can display a checkbox before the fields in order to specify if the configuration value is set for the current context shop/group,  or if it is a value inherited from all contexts, or from a shop group.

Reviewer: ```suggestion
Since Prestashop 1.7.8, when multistore is enabled, migrated configuration forms can display a checkbox before the fields in order to specify if the configuration value is set for the current context shop/group, or if it is a value inherited from all contexts, or from a shop group.
```

```php
<?php

    /**
     * {@inheritdoc}
     *
     * @see MultistoreConfigurationTypeExtension
     */
    public function getParent(): string
    {
        return MultistoreConfigurationType::class;
    }
```

Then for each form elements that should have its multistore checkbox, you must add a `multistore_configuration_key` attribute, with the configuration key to which it should be attached as a value:

```php
<?php
class MyFormType extends TranslatorAwareType
{
    /**
     * {@inheritdoc}
     */
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder
            ->add(
                'my_field',
                SwitchType::class,
                [
                    'required' => true,
                    // here we add the multistore_configuration_key attribute with the configuration key it is linked to.
                    'attr' => [
                        'multistore_configuration_key' => 'THE_CONFIGURATION_KEY',
                    ],
                    // ...
                ]
            )
        // ...

```

Doing this will have several effects:

> - Each field implementing the `multistore_configuration_key` will have a multistore checkbox before them.
> - The checkboxes will automatically be checked or not depending on the configuration for the current shop context.
> - When the configuration value is inherited, the checkbox will be unchecked, and the field disabled and greyed out.

Reviewer: ```suggestion
- Each field implementing the `multistore_configuration_key` will have a multistore checkbox before them ;
- The checkboxes will automatically be checked or not depending on the configuration for the current shop context ;
- When the configuration value is inherited, the checkbox will be unchecked, and the field disabled and greyed out ;
```

Author: Is there a reason for replacing the final dot with a `;`?

Reviewer 2: yes, it's common to use ; or , in list and start every point with small letter :D

> - Checking or unchecking a checkbox will enable/disable the field for the current context.

Reviewer 2: ```suggestion
- Checking or unchecking a checkbox will enable/disable the field for the current shop context.
```

> ## Saving data from a multistore configuration form

Reviewer 2: I know it's not part of the docs, but this code seems to be very generic, can we have it globally available instead of need to repeat it in every controller?

Author: That's a good idea, I'll make a modification in this direction on the PR and then modify this PR

Reviewer 2: awesome 💪🏻

> - When receiving the data from the form, you will also receive the values from the multistore checkboxes that were added to your form.

Reviewer: ```suggestion
- When receiving the data from the form, you will also receive values from the multistore checkboxes that were added to your form.
```

- Fields having their multistore checkboxes unchecked should not be saved.

> Multistore checkboxes' names are composed of `multistore_` followed by the name of the field it is applied to. For example the checkbox having the name `multistore_maintenance_text` applies to the field having the name `maintenance_text`. On the frontend part, if the checkbox is unchecked, the corresponding field has been disabled, but it would be a good idea to check in the backend that when a checkbox is unchecked, you don't deal with the data of its corresponding field, and to remove the disabled fields from the list of fields that will be treated, here is an example from `src/Adapter/Shop/MaintenanceConfiguration.php`:

Reviewer: ```suggestion
Multistore checkboxes' names began with `multistore_` followed by the name of the field it is applied to. For example the checkbox having the name `multistore_maintenance_text` applies to the field having the name `maintenance_text`. On the frontend part, if the checkbox is unchecked, the corresponding field has been disabled, but it would be a good idea to check in the backend that when a checkbox is unchecked, you don't deal with the data of its corresponding field, and to remove the disabled fields from the list of fields that will be treated, here is an example from `src/Adapter/Shop/MaintenanceConfiguration.php`:
```

```php
<?php

    public function removeDisabledFields(array $configuration): array
    {
        if ($this->shopContext->isAllShopContext()) {
            return $configuration;
        }

>         // remove fields that have their multistore checkbox unchecked

Reviewer: ```suggestion
        // Remove fields that have their multistore checkbox unchecked
```

---
title: Configuration forms
menuTitle: Configuration forms
summary: "Configuration forms when multistore is enabled"
weight: 50
---

# Configuration forms in multistore context
{{< minver v="1.7.8" title="true" >}}

> Since Prestashop 1.7.8, when multistore is enabled, migrated configuration forms can display a checkbox before the fields in order to specify if the configuration value is set for the current context shop/group, or if it is a value inherited from all contexts, or from a shop group.

Reviewer: ```suggestion
Since Prestashop 1.7.8, when multistore is enabled, migrated configuration forms can display a checkbox before fields in order to specify if the configuration value is set for the current context shop/group, or if it is a value inherited from all contexts, or from a shop group.
```

> If you're developing a configuration form and you want it to take into account the current multistore context, then you must make your form compatible with our multistore checkboxes. This is made possible thanks to [Symfony's form extensions mechanism](https://symfony.com/doc/3.4/form/create_form_type_extension.html).

Reviewer: ```suggestion
If you're developing a configuration form and you want it to take into account the current multistore context, then you must make your form compatible with multistore checkboxes. This is made possible thanks to [Symfony's form extensions mechanism](https://symfony.com/doc/3.4/form/create_form_type_extension.html).
```

- When a multistore checkbox is pushed checked, we want to save the configuration value for the current context.

When your configuration form is multistore compatible, your configuration class should extend the `AbstractMultistoreConfiguration` abstract class instead of directly implementing the `DataConfigurationInterface` interface. This will give you access to helper methods that will make it easy to store the right configuration values at the right place and for the right context.

The two helper methods are

- **`getShopConstraint()`:** this will give you a ShopConstraint object reflecting the current context
```php
public function getShopConstraint(): ?ShopConstraint
```
- **`updateConfigurationValue()`:** this will save or delete the configuration value for the given configuration key, taking into account the state of the multistore checkbox and the current shop context. Note that this helper method is totally compatible with non multistore fields.
```php
public function updateConfigurationValue(string $configurationKey, string $fieldName, array $input, ?ShopConstraint $shopConstraint, array $options = []): void
```

> As an example, see how it's used in the `MaintenanceConfiguration` class

Reviewer: Add the namespace behind the class (for easily find it).
